File: 4_Behavioral_Design_Patterns/4.04_Iterator/Python_Snippet/Iterator.py
# Book class representing a book
class Book:
    def __init__(self, title):
        self.title = title

    def get_title(self):
        return self.title

# Library class representing a collection of books
class Library:
    def __init__(self):
        self.books = []

    def add_book(self, book):
        self.books.append(book)

    # Iterator class to iterate over the collection of books
    class BookIterator:
        def __init__(self, library):
            self.library = library
            self.current_index = 0

        def __iter__(self):
            return self

        def __next__(self):
            if self.current_index < len(self.library.books):
                book = self.library.books[self.current_index]
                self.current_index += 1
                return book
            else:
                raise StopIteration

    # Method to get an iterator for the collection of books
    def __iter__(self):
        return self.BookIterator(self)

# Main function
def main():
    library = Library()
    library.add_book(Book("Book 1"))
    library.add_book(Book("Book 2"))
    library.add_book(Book("Book 3"))

    # Iterate over the collection of books using the Iterator
    for book in library:
        print("Book Title:", book.get_title())

if __name__ == "__main__":
    main()

File: 4_Behavioral_Design_Patterns/4.04_Iterator/Java_Snippet/Iterator.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

// Book class representing a book
class Book {
    private String title;

    public Book(String title) {
        this.title = title;
    }

    public String getTitle() {
        return title;
    }
}

// Library class representing a collection of books
class Library {
    private List<Book> books;

    public Library() {
        this.books = new ArrayList<>();
    }

    public void addBook(Book book) {
        books.add(book);
    }

    // Iterator class to iterate over the collection of books
    private class BookIterator implements Iterator<Book> {
        private int currentIndex = 0;

        @Override
        public boolean hasNext() {
            return currentIndex < books.size();
        }

        @Override
        public Book next() {
            return books.get(currentIndex++);
        }
    }

    // Method to get an iterator for the collection of books
    public Iterator<Book> iterator() {
        return new BookIterator();
    }
}

public class Main {
    public static void main(String[] args) {
        Library library = new Library();
        library.addBook(new Book("Book 1"));
        library.addBook(new Book("Book 2"));
        library.addBook(new Book("Book 3"));

        // Iterate over the collection of books using the Iterator
        Iterator<Book> iterator = library.iterator();
        while (iterator.hasNext()) {
            Book book = iterator.next();
            System.out.println("Book Title: " + book.getTitle());
        }
    }
}
